### PR TITLE
Enhancement: Reducing memory usage by removing the vtable in managed.

### PR DESCRIPTION
**Pull request by RealTimeChris**
Have applied the CRTP to keep the interfaces identical, as well as added snowflakes to the necessary classes, and also added a managed_base class to allow for storing them all in the same deletion_queue.

## Code change checklist

- [x] I have ensured that all methods and functions are **fully documented** using doxygen style comments
- [x] My code follows the [coding style guide](https://dpp.dev/coding-standards.html).
- [x] I tested that my change works before raising the PR.
- [x] I have ensured that I did not break any existing API calls.
- [x] I have not built my pull request using AI, a static analysis tool or similar without any human oversight.

### CODE REVIEW

File: include/dpp/integration.h [hunk 1]
```diff
@@ -80,8 +80,10 @@ struct DPP_EXPORT integration_app {
 /**
  * @brief Represents an integration on a guild, e.g. a connection to twitch.
  */
-class DPP_EXPORT integration : public managed, public json_interface<integration> {
+class DPP_EXPORT integration : public managed<integration>, public json_interface<integration> {
 public:
+	//** Unique identifier */
+	snowflake id;
```
Comment (braindigitalis) [inline include/dpp/integration.h:86]: I don't really like this, it's basically added tons of duplicate lines to every object to save 8 bytes of ram per instance, I think readability is more important in this case

**Comment by Mishura4** [comment]
The id could be part of managed<T>, that would help with the code duplication

**Comment by RealTimeChris** [comment]
> The id could be part of managed, that would help with the code duplication

It was deliberately removed from managed in order to eliminate the vtable.

**Comment by Mishura4** [comment]
> > The id could be part of managed, that would help with the code duplication
> 
> It was deliberately removed from managed in order to eliminate the vtable.

Having just the snowflake in there isn't going to produce a vtable though.

**Comment by RealTimeChris** [comment]
> > > The id could be part of managed, that would help with the code duplication
> > 
> > 
> > It was deliberately removed from managed in order to eliminate the vtable.
> 
> Having just the snowflake in there isn't going to produce a vtable though.

No but having the virtual destructor that is then required to properly destroy the id will.

**Comment by Mishura4** [comment]
Virtual destructor is needed if you want to destroy a variable through a pointer to its base class, so the derived object is properly destroyed. Whether the snowflake is in the base or not doesn't change that

**Pull request closed without merging**